ocaml/xapi/xapi_xenops.ml

```diff
@@ -2033,6 +2033,20 @@ let update_vm ~__context id =
                   "Will update VM.allowed_operations because power_state has \
                    changed." ;
                 should_update_allowed_operations := true ;
+                (* Update ha_always_run before the power_state (if needed), to avoid racing
```

Comment:
so the ha monitor thread checks the field `ha_always_run` to determine whether it needs to restart a protected VM. Is it possible to add this logic into `Helpers.vm_should_always_run` (although this one does not have `context`) or `xapi_ha_vm_failover.all_protected_vms`?

Reply:
Should the documentation of the field in the IDL reference  `ha_always_run` as a dependency?

Reply:
@lindig `VM.ha_always_run` is really an internal implementation detail now and API clients shouldn't really look at it anymore. Initially, this field was the way to control protected VMs using the API, but it has been deprecated for a long time.

Reply:
@Vincent-lau I think it will be hard to encode this in some sort of new state without catching the shutdown event here.